File: 400.NthDigit/nDigit.py
class Solution(object):
    def findNthDigit(self, n):
        """
        :type n: int
        :rtype: int
        """
        k = base = 0
        while True:
            tmp = int(9 * k * (10 ** (k-1)))
            if base + tmp > n:
                break
            base += tmp
            k += 1

        remain = n - base
        num = -(-remain / k) + int(10 ** (k-1)) - 1
        nDigit = -remain % k
        print num, nDigit, k

        for i in range(nDigit+1):
            result = num % 10
            num /= 10

        return result


if __name__ == '__main__':
    print Solution().findNthDigit(1000000000)

File: 400.NthDigit/nDigit.cc
class Solution {
public:
    int findNthDigit(int n) {
        int k = 0, base = 0;
        while (true) {
            long tmp = 9 * k * (long)pow(10, k-1);
            if (tmp > (long)n) {
                break;
            }
            base += tmp;
            k++;
        }

        int remain = n - base;
        int num = (int)pow(10, k-1) - 1 + (remain / k + (remain % k > 0 ? 1 : 0));
        int nDigit = (k - remain % k) % k;

        int ans = 0;
        for (int i = 0; i < nDigit + 1; i++) {
            ans = num % 10;
            num /= 10;
        }

        return ans;
    }
};
